jwst/dark_current/dark_class.py

```diff
@@ -0,0 +1,73 @@
+import numpy as np
+
+
+class DarkData:
+    """
+    Class removing data model dependencies.
+    """
+    def __init__(self, dims=None, dark_model=None):
+        """
+        Creates a class to remove data model dependencies in the internals of
+        the dark current code.
+
+        Paramters
+        ---------
+        dims: tuple
+            A tuple of integers to describe the dimensions of the arrays used
+            during the dark current step.
+
+        dark_model: data model
+            Input data model, assumed to be a JWST DarkModel like model.
+        """
+        if dark_model is not None:
+            self.data = dark_model.data
+            self.dq = dark_model.dq
+            self.err = dark_model.err
+
+            self.exp_nframes = dark_model.meta.exposure.nframes
+            self.exp_ngroups = dark_model.meta.exposure.ngroups
+            self.exp_groupgap = dark_model.meta.exposure.groupgap
+
+        elif dims is not None:
+            self.data = np.zeros(dims, dtype=np.float32)
+            self.dq = np.zeros(dims, dtype=np.uint8)
+            self.err = np.zeros(dims, dtype=np.float32)
+
+            self.exp_nframes = None
+            self.exp_ngroups = None
+            self.exp_groupgap = None
+
+        else:
+            self.data = None
+            self.dq = None
+            self.err = None
+
+            self.exp_nframes = None
+            self.exp_ngroups = None
+            self.exp_groupgap = None
+
+        self.save = False
+        self.oname = None
+
+
+class DarkScienceData:
+    def __init__(self, science_model):
+        """
+        Creates a class to remove data model dependencies in the internals of
+        the dark current code.
+
+        Paramters
+        ---------
+        science_model: data model
+            Input data model, assumed to be a JWST RampModel like model.
+        """
+        self.data = science_model.data
+        self.dq = science_model.groupdq
```

Comment:
The science dimensions is something we need to talk about.  Since Roman science arrays will be 3-D and JWST will be 4-D, some handling somewhere needs to be added for this.  I know for ramp fitting, the Roman step code will simply add a fourth dimension as 1 to make (ngroups, nrows, ncols) into (1, ngroups, nrows, ncols).  This prevents internal refactoring of ramp fitting code.  

For dark current, there are some internal handling of dimensions based on instrument type.  Is this dimension handling the same for Roman as it is for JWST?  Is the above ramp fitting work around a valid way to work around the dimensions for dark current for the different telescope pipelines?

Reply:
For JWST the input science data in the dark_current step are *always* 4-D (nints, ngroups, nrows, ncols). What can vary is the shape of the dark *reference file* data. For most JWST instruments, the dark ref file data are 3-D (ngroups, nrows, ncols) and are subtracted from the 4-D science data by just broadcasting to the 4th dimension (i.e. same correction is applied to all data along the 4th dimension). For MIRI, however, the dark correction is somewhat dependent on integration and hence the MIRI dark reference file data are 4-D. That's the special handling that you see going on down inside of the step. Since Roman data will always be only 3-D for both science and dark reference file, it shouldn't need any of this special handling.

Reply:
@hbushouse if roman is not going to have separate pixel/group dq arrays, then many of the existing steps in stcal will have to be modified because they currently require both of these arrays as inputs. is this something that should be fixed and considered going forward?

Reply:
I don't know what the plans are for Roman datamodels, in terms of whether they'll use the same kind of `groupdq` and `pixeldq` arrays as JWST datamodels. Maybe @sosey or @nden know?  You're right that it would require quite a bit of retooling of lower-level functions in many steps if they were different from JWST.